From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Baocheng Su <baocheng.su@siemens.com>
Date: Mon, 6 Jul 2026 19:55:19 +0800
Subject: [PATCH] gpio: add GPIO character device v2 support

Add GPIO character device v2 support, including line discovery,
line requests, line reconfiguration, and GPIO value access.

Track the character device API version in platform metadata and
integrate v2 line groups into the common GPIO initialization and
cleanup paths while preserving compatibility with the v1 interface.

Signed-off-by: Baocheng Su <baocheng.su@siemens.com>
---
 include/gpio/gpio_chardev_v2.h        |  31 ++
 include/linux/gpio.h                  |  98 ++++
 include/mraa_internal_types.h         |   7 +-
 src/CMakeLists.txt                    |   1 +
 src/arm/96boards.c                    |  12 +-
 src/arm/rockpi4.c                     |   2 +-
 src/gpio/gpio.c                       | 717 ++++++++++++++++++--------
 src/gpio/gpio_chardev.c               |   4 +
 src/gpio/gpio_chardev_v2.c            | 185 +++++++
 src/mraa.c                            |  12 +-
 src/x86/adlink-ipi.c                  |   2 +-
 src/x86/iei_tank.c                    |   2 +-
 src/x86/intel_ilk.c                   |   2 +-
 src/x86/intel_joule_expansion.c       |   2 +-
 src/x86/intel_minnow_byt_compatible.c |   2 +-
 src/x86/up2.c                         |   2 +-
 src/x86/up_xtreme.c                   |   2 +-
 17 files changed, 858 insertions(+), 225 deletions(-)
 create mode 100644 include/gpio/gpio_chardev_v2.h
 create mode 100644 src/gpio/gpio_chardev_v2.c

diff --git a/include/gpio/gpio_chardev_v2.h b/include/gpio/gpio_chardev_v2.h
new file mode 100644
index 000000000000..74e1d80bf445
--- /dev/null
+++ b/include/gpio/gpio_chardev_v2.h
@@ -0,0 +1,31 @@
+/*
+ * Author: Baocheng Su <baocheng.su@siemens.com>
+ * Copyright (c) 2026 Siemens AG
+ *
+ * SPDX-License-Identifier: MIT
+ */
+
+#pragma once
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#include "gpio_chardev.h"
+#include "linux/gpio.h"
+#include "mraa_internal.h"
+
+int mraa_gpiod_get_lines_request(int chip_fd, unsigned int line_offsets[], unsigned num_lines);
+int mraa_gpiod_get_line_name(int chip_fd, unsigned line_offset, char* name, size_t name_size);
+int mraa_gpiod_get_line_flags(int chip_fd, unsigned line_offset, unsigned long* flag);
+
+int mraa_gpiod_find_gpio_line_by_name(const char* name, unsigned* chip_number, unsigned* line_number);
+
+int mraa_gpiod_reconfig_lines(int line_req_fd, unsigned long flag);
+
+int mraa_gpiod_get_values(int line_req_fd, unsigned num_lines, unsigned char line_values[]);
+int mraa_gpiod_set_values(int line_req_fd, unsigned num_lines, unsigned char line_values[]);
+
+#ifdef __cplusplus
+}
+#endif
diff --git a/include/linux/gpio.h b/include/linux/gpio.h
index 8451032c6703..452a42058ceb 100644
--- a/include/linux/gpio.h
+++ b/include/linux/gpio.h
@@ -78,4 +78,102 @@ struct gpioevent_data {
 #define GPIO_GET_LINEHANDLE_IOCTL _IOWR(0xB4, 0x03, struct gpiohandle_request)
 #define GPIO_GET_LINEEVENT_IOCTL _IOWR(0xB4, 0x04, struct gpioevent_request)
 
+/**
+ * V2 API
+ */
+#define GPIO_V2_LINES_MAX 64
+#define GPIO_MAX_NAME_SIZE 32
+#define GPIO_V2_LINE_NUM_ATTRS_MAX 10
+
+enum gpio_v2_line_flag {
+    GPIO_V2_LINE_FLAG_USED = 1 << 0,
+    GPIO_V2_LINE_FLAG_ACTIVE_LOW = 1 << 1,
+    GPIO_V2_LINE_FLAG_INPUT = 1 << 2,
+    GPIO_V2_LINE_FLAG_OUTPUT = 1 << 3,
+    GPIO_V2_LINE_FLAG_EDGE_RISING = 1 << 4,
+    GPIO_V2_LINE_FLAG_EDGE_FALLING = 1 << 5,
+    GPIO_V2_LINE_FLAG_OPEN_DRAIN = 1 << 6,
+    GPIO_V2_LINE_FLAG_OPEN_SOURCE = 1 << 7,
+    GPIO_V2_LINE_FLAG_BIAS_PULL_UP = 1 << 8,
+    GPIO_V2_LINE_FLAG_BIAS_PULL_DOWN = 1 << 9,
+    GPIO_V2_LINE_FLAG_BIAS_DISABLED = 1 << 10,
+    GPIO_V2_LINE_FLAG_EVENT_CLOCK_REALTIME = 1 << 11,
+    GPIO_V2_LINE_FLAG_EVENT_CLOCK_HTE = 1 << 12,
+};
+
+struct gpio_v2_line_attribute {
+	__u32 id;
+	__u32 padding;
+	union {
+		__aligned_u64 flags;
+		__aligned_u64 values;
+		__u32 debounce_period_us;
+	};
+};
+
+struct gpio_v2_line_config_attribute {
+	struct gpio_v2_line_attribute attr;
+	__aligned_u64 mask;
+};
+
+struct gpio_v2_line_config {
+	__aligned_u64 flags;
+	__u32 num_attrs;
+	/* Pad to fill implicit padding and reserve space for future use. */
+	__u32 padding[5];
+	struct gpio_v2_line_config_attribute attrs[GPIO_V2_LINE_NUM_ATTRS_MAX];
+};
+
+struct gpio_v2_line_request {
+	__u32 offsets[GPIO_V2_LINES_MAX];
+	char consumer[GPIO_MAX_NAME_SIZE];
+	struct gpio_v2_line_config config;
+	__u32 num_lines;
+	__u32 event_buffer_size;
+	/* Pad to fill implicit padding and reserve space for future use. */
+	__u32 padding[5];
+	__s32 fd;
+};
+
+struct gpio_v2_line_info {
+    char name[GPIO_MAX_NAME_SIZE];
+    char consumer[GPIO_MAX_NAME_SIZE];
+    __u32 offset;
+    __u32 num_attrs;
+    __aligned_u64 flags;
+    struct gpio_v2_line_attribute attrs[GPIO_V2_LINE_NUM_ATTRS_MAX];
+    /* Space reserved for future use. */
+    __u32 padding[4];
+};
+
+struct gpio_v2_line_values {
+    __aligned_u64 bits;
+    __aligned_u64 mask;
+};
+
+enum gpio_v2_line_event_id {
+    GPIO_V2_LINE_EVENT_RISING_EDGE = 1,
+    GPIO_V2_LINE_EVENT_FALLING_EDGE = 2,
+};
+
+struct gpio_v2_line_event {
+    __aligned_u64 timestamp_ns;
+    __u32 id;
+    __u32 offset;
+    __u32 seqno;
+    __u32 line_seqno;
+    /* Space reserved for future use. */
+    __u32 padding[6];
+};
+
+/*
+ * v2 ioctl()s
+ */
+#define GPIO_V2_GET_LINEINFO_IOCTL _IOWR(0xB4, 0x05, struct gpio_v2_line_info)
+#define GPIO_V2_GET_LINEINFO_WATCH_IOCTL _IOWR(0xB4, 0x06, struct gpio_v2_line_info)
+#define GPIO_V2_GET_LINE_IOCTL _IOWR(0xB4, 0x07, struct gpio_v2_line_request)
+#define GPIO_V2_LINE_SET_CONFIG_IOCTL _IOWR(0xB4, 0x0D, struct gpio_v2_line_config)
+#define GPIO_V2_LINE_GET_VALUES_IOCTL _IOWR(0xB4, 0x0E, struct gpio_v2_line_values)
+#define GPIO_V2_LINE_SET_VALUES_IOCTL _IOWR(0xB4, 0x0F, struct gpio_v2_line_values)
+
 #endif /* _GPIO_H_ */
diff --git a/include/mraa_internal_types.h b/include/mraa_internal_types.h
index 63f2b94fef2f..52634a018f91 100644
--- a/include/mraa_internal_types.h
+++ b/include/mraa_internal_types.h
@@ -94,7 +94,10 @@ struct _firmata {
 struct _gpio_group {
     int is_required;
     int dev_fd;
-    int gpiod_handle;
+    union {
+        int gpiod_handle;
+        int line_request_fd;
+    };
     unsigned int gpio_chip;
     /* We can have multiple lines in a gpio group. */
     unsigned int num_gpio_lines;
@@ -499,7 +502,7 @@ typedef struct _board_t {
     mraa_pininfo_t* pins;     /**< Pointer to pin array */
     mraa_adv_func_t* adv_func;    /**< Pointer to advanced function disptach table */
     struct _board_t* sub_platform;     /**< Pointer to sub platform */
-    mraa_boolean_t chardev_capable;  /**< Decide what interface is being used: old sysfs or new char device*/
+    unsigned int chardev_ver;  /**< Decide what interface is being used: 0 for old sysfs, 1 for chardev V1, 2 for chardev V2 */
     mraa_led_dev_t led_dev[MAX_LED_COUNT]; /**< Array of LED devices */
     unsigned int led_dev_count; /**< Total onboard LED device count */
     /*@}*/
diff --git a/src/CMakeLists.txt b/src/CMakeLists.txt
index 8146eb674ef9..5c4922ff283a 100644
--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -24,6 +24,7 @@ set (mraa_LIB_SRCS_NOAUTO
   ${PROJECT_SOURCE_DIR}/src/mraa.c
   ${PROJECT_SOURCE_DIR}/src/gpio/gpio.c
   ${PROJECT_SOURCE_DIR}/src/gpio/gpio_chardev.c
+  ${PROJECT_SOURCE_DIR}/src/gpio/gpio_chardev_v2.c
   ${PROJECT_SOURCE_DIR}/src/i2c/i2c.c
   ${PROJECT_SOURCE_DIR}/src/pwm/pwm.c
   ${PROJECT_SOURCE_DIR}/src/spi/spi.c
diff --git a/src/arm/96boards.c b/src/arm/96boards.c
index ebaabae9690a..95c456fbbb88 100644
--- a/src/arm/96boards.c
+++ b/src/arm/96boards.c
@@ -266,41 +266,41 @@ mraa_96boards()
             b->led_dev[5].name = (char*) db410c_led[5];
             b->led_dev_count = MRAA_96BOARDS_LED_COUNT;
             b->adv_func->gpio_mmap_setup = &mraa_db410c_mmap_setup;
-            b->chardev_capable = 1;
+            b->chardev_ver = 1;
         } else if (mraa_file_contains(DT_BASE "/model", "Qualcomm Technologies, Inc. DB820c")) {
             b->platform_name = PLATFORM_NAME_DB820C;
             ls_gpio_pins = db820c_ls_gpio_pins;
             chardev_map = &db820c_chardev_map;
             b->uart_dev[0].device_path = (char*) db820c_serialdev[0];
             b->uart_dev[1].device_path = (char*) db820c_serialdev[1];
-            b->chardev_capable = 1;
+            b->chardev_ver = 1;
         } else if (mraa_file_contains(DT_BASE "/model", "HiKey Development Board")) {
             b->platform_name = PLATFORM_NAME_HIKEY;
             ls_gpio_pins = hikey_ls_gpio_pins;
             chardev_map = &hikey_chardev_map;
             b->uart_dev[0].device_path = (char*) hikey_serialdev[0];
             b->uart_dev[1].device_path = (char*) hikey_serialdev[1];
-            b->chardev_capable = 1;
+            b->chardev_ver = 1;
         } else if (mraa_file_contains(DT_BASE "/model", "HiKey960")) {
             b->platform_name = PLATFORM_NAME_HIKEY960;
             chardev_map = &hikey960_chardev_map;
             b->uart_dev[0].device_path = (char*) hikey960_serialdev[0];
             b->uart_dev[1].device_path = (char*) hikey960_serialdev[1];
-            b->chardev_capable = 1;
+            b->chardev_ver = 1;
         } else if (mraa_file_contains(DT_BASE "/model", "ROCK960")) {
             b->platform_name = PLATFORM_NAME_ROCK960;
             ls_gpio_pins = rock960_ls_gpio_pins;
             chardev_map = &rock960_chardev_map;
             b->uart_dev[0].device_path = (char*) rock960_serialdev[0];
             b->uart_dev[1].device_path = (char*) rock960_serialdev[1];
-            b->chardev_capable = 1;
+            b->chardev_ver = 1;
         } else if ((mraa_file_contains(DT_BASE "/model", "ZynqMP ZCU100 RevC")) ||
                    (mraa_file_contains(DT_BASE "/model", "Avnet Ultra96 Rev1"))) {
             b->platform_name = PLATFORM_NAME_ULTRA96;
             chardev_map = &ultra96_chardev_map;
             b->uart_dev[0].device_path = (char*) ultra96_serialdev[0];
             b->uart_dev[1].device_path = (char*) ultra96_serialdev[1];
-            b->chardev_capable = 1;
+            b->chardev_ver = 1;
         }
     }
 
diff --git a/src/arm/rockpi4.c b/src/arm/rockpi4.c
index dfa39a85ab21..7499521288bb 100644
--- a/src/arm/rockpi4.c
+++ b/src/arm/rockpi4.c
@@ -123,7 +123,7 @@ mraa_rockpi4()
     b->adc_supported = 10;
     b->aio_dev[0].pin = 26;
     b->aio_non_seq = 1;
-    b->chardev_capable = 1;
+    b->chardev_ver = 1;
 
     mraa_rockpi4_pininfo(b, 0,   -1, (mraa_pincapabilities_t){0,0,0,0,0,0,0,0}, "INVALID");
     mraa_rockpi4_pininfo(b, 1,   -1, (mraa_pincapabilities_t){1,0,0,0,0,0,0,0}, "3V3");
diff --git a/src/gpio/gpio.c b/src/gpio/gpio.c
index 5ea1c4b53a91..7708a3c4e4e3 100644
--- a/src/gpio/gpio.c
+++ b/src/gpio/gpio.c
@@ -7,6 +7,7 @@
  */
 #include "gpio.h"
 #include "gpio/gpio_chardev.h"
+#include "gpio/gpio_chardev_v2.h"
 #include "linux/gpio.h"
 #include "mraa_internal.h"
 
@@ -50,14 +51,17 @@ mraa_gpio_close_event_handles_sysfs(int fds[], int num_fds)
         return;
     }
 
-    for (int i = 0; i < num_fds; ++i) {
-        // Check required to avoid closing stdin and of an uninitialized fd
-        if (fds[i] != 0) {
-            close(fds[i]);
+    if (plat->chardev_ver != 2) {
+        // Do nothing for V2, as the line_request_fd is closed in _mraa_free_gpio_groups
+        for (int i = 0; i < num_fds; ++i) {
+            // Check required to avoid closing stdin and of an uninitialized fd
+            if (fds[i] != 0) {
+                close(fds[i]);
+            }
         }
-    }
 
-    free(fds);
+        free(fds);
+    }
 }
 
 static mraa_gpio_context
@@ -100,7 +104,7 @@ mraa_gpio_init_internal(mraa_adv_func_t* func_table, int pin)
     dev->isr_thread_terminating = 0;
     dev->phy_pin = -1;
 
-    if ((plat != NULL) && (!plat->chardev_capable)) {
+    if ((plat != NULL) && (!plat->chardev_ver)) {
         char bu[MAX_SIZE];
         int length;
 
@@ -149,24 +153,46 @@ init_internal_cleanup:
 mraa_gpio_context
 mraa_gpio_init_by_name(char* name)
 {
-    mraa_board_t* board = plat;
+    unsigned int chardev_api_ver = plat->chardev_ver;
+    unsigned chip_number, line_offset;
     mraa_gpio_context dev;
-    mraa_gpiod_group_t gpio_group;
-    mraa_gpiod_line_info* linfo = NULL;
-    mraa_gpiod_chip_info* cinfo;
-    mraa_gpiod_chip_info** cinfos = NULL;
-    int i, line_found, line_offset, chip_fd;
+    int chip_fd, ret;
 
     if (name == NULL) {
         syslog(LOG_ERR, "[GPIOD_INTERFACE]: Gpio name not valid");
         return NULL;
     }
 
-    if (!board->chardev_capable) {
+    if (!chardev_api_ver) {
         syslog(LOG_ERR, "[GPIOD_INTERFACE]: gpio_init_by_name not available for this platform!");
         return NULL;
     }
 
+    switch (chardev_api_ver) {
+        case 1:
+            ret = mraa_find_gpio_line_by_name(name, &chip_number, &line_offset);
+            break;
+        case 2:
+            ret = mraa_gpiod_find_gpio_line_by_name(name, &chip_number, &line_offset);
+            break;
+        default:
+            syslog(LOG_ERR, "[GPIOD_INTERFACE]: Invalid chardev_api_ver value: %d", chardev_api_ver);
+            return NULL;
+    }
+
+    if (ret < 0) {
+        syslog(LOG_ERR, "[GPIOD_INTERFACE]: Gpio name not found");
+        return NULL;
+    }
+
+    mraa_gpiod_chip_info* cinfo = mraa_get_chip_info_by_number(chip_number);
+    if (cinfo == NULL) {
+        syslog(LOG_ERR, "[GPIOD_INTERFACE]: Failed to get chip info for chip number %u", chip_number);
+        return NULL;
+    }
+    chip_fd = cinfo->chip_fd;
+    free(cinfo);
+
     dev = (mraa_gpio_context) calloc(1, sizeof(struct _gpio));
     if (dev == NULL) {
         syslog(LOG_CRIT, "[GPIOD_INTERFACE]: Failed to allocate memory for context");
@@ -180,86 +206,51 @@ mraa_gpio_init_by_name(char* name)
         return NULL;
     }
 
-    dev->num_chips = mraa_get_chip_infos(&cinfos);
-    if (dev->num_chips <= 0) {
-        mraa_gpio_close(dev);
-        return NULL;
-    }
-
     /* We are dealing with a single GPIO */
+    dev->num_chips = 1;
     dev->num_pins = 1;
 
-    gpio_group = calloc(dev->num_chips, sizeof(struct _gpio_group));
-    if (gpio_group == NULL) {
+    dev->gpio_group = calloc(1, sizeof(struct _gpio_group));
+    if (dev->gpio_group == NULL) {
         syslog(LOG_CRIT, "[GPIOD_INTERFACE]: Failed to allocate memory for internal member");
-        for_each_gpio_chip(cinfo, cinfos, dev->num_chips) {
-            if (cinfo) close(cinfo->chip_fd);
-        }
-        free(cinfos);
         mraa_gpio_close(dev);
         return NULL;
     }
 
-    dev->gpio_group = gpio_group;
-    for (i = 0; i < dev->num_chips; ++i) {
-        gpio_group[i].gpio_chip = i;
-        gpio_group[i].gpio_lines = NULL;
-    }
-
-    /* Iterate over all gpiochips in the platform to find the requested line */
-    for_each_gpio_chip(cinfo, cinfos, dev->num_chips)
-    {
-        for (i = 0; i < cinfo->chip_info.lines; i++) {
-            linfo = mraa_get_line_info_by_chip_name(cinfo->chip_info.name, i);
-            if (!strncmp(linfo->name, name, 32)) {
-                /* idx is coming from `for_each_gpio_chip` definition */
-                syslog(LOG_DEBUG, "[GPIOD_INTERFACE]: Chip: %d Line: %d", idx, i);
-                if (!gpio_group[idx].is_required) {
-                    gpio_group[idx].dev_fd = cinfo->chip_fd;
-                    gpio_group[idx].is_required = 1;
-                    gpio_group[idx].gpiod_handle = -1;
-                    chip_fd = gpio_group[idx].dev_fd;
-                }
-
-                /* Map pin to _gpio_group structure. */
-                dev->pin_to_gpio_table[0] = idx;
-                gpio_group[idx].gpio_lines = realloc(gpio_group[idx].gpio_lines, sizeof(unsigned int));
-                gpio_group[idx].gpio_lines[0] = i;
-                gpio_group[idx].num_gpio_lines++;
-
-                line_found = 1;
-                line_offset = i;
-
-                free(linfo);
-                break;
-            }
-            free(linfo);
-        }
-    }
-
-    for_each_gpio_chip(cinfo, cinfos, dev->num_chips) {
-        if (!line_found || cinfo->chip_fd != chip_fd) {
-            close(cinfo->chip_fd);
-        }
-    }
-    free(cinfos);
-
-    if (!line_found) {
-        syslog(LOG_ERR, "[GPIOD_INTERFACE]: Gpio not found!");
+    dev->gpio_group->is_required = 1;
+    dev->gpio_group->dev_fd = chip_fd;
+    dev->gpio_group->gpio_chip = chip_number;
+    dev->gpio_group->gpio_lines = calloc(1, sizeof(unsigned int));
+    if (dev->gpio_group->gpio_lines == NULL) {
+        syslog(LOG_CRIT, "[GPIOD_INTERFACE]: Failed to allocate memory for internal member");
         mraa_gpio_close(dev);
         return NULL;
     }
+    *dev->gpio_group->gpio_lines = line_offset;
+    dev->gpio_group->num_gpio_lines = 1;
 
-    /* Initialize rw_values for read / write multiple functions */
-    for (i = 0; i < dev->num_chips; ++i) {
-        gpio_group[i].rw_values = calloc(gpio_group[i].num_gpio_lines, sizeof(unsigned char));
-        if (gpio_group[i].rw_values == NULL) {
-            syslog(LOG_CRIT, "[GPIOD_INTERFACE]: Failed to allocate memory for internal member");
+    if (chardev_api_ver == 1) {
+        dev->gpio_group->gpiod_handle = -1;
+        dev->gpio_group->event_handles = NULL;
+    } else {
+        int line_request_fd = mraa_gpiod_get_lines_request(chip_fd, dev->gpio_group->gpio_lines, 1);
+        if (line_request_fd < 0) {
+            syslog(LOG_ERR, "[GPIOD_INTERFACE]: error requesting lines for chip %d", chip_number);
             mraa_gpio_close(dev);
             return NULL;
         }
 
-        gpio_group[i].event_handles = NULL;
+        dev->gpio_group->line_request_fd = line_request_fd;
+    }
+
+    dev->pin_to_gpio_table[0] = 0; /* Map the single pin to the first (and only) gpio group */
+
+    /* Initialize rw_values for read / write multiple functions */
+    dev->gpio_group->rw_values = calloc(1, sizeof(unsigned char));
+    if (dev->gpio_group->rw_values == NULL) {
+        syslog(LOG_CRIT, "[GPIOD_INTERFACE]: Failed to allocate memory for internal member");
+        mraa_gpio_close(dev);
+        return NULL;
     }
 
     /* Save the provided array from the user to our internal structure. */
@@ -305,7 +296,7 @@ mraa_gpio_init(int pin)
         pin = mraa_get_sub_platform_index(pin);
     }
 
-    if (board->chardev_capable) {
+    if (board->chardev_ver) {
         int pins[1] = { pin };
         return mraa_gpio_init_multi(pins, 1);
     }
@@ -351,7 +342,7 @@ mraa_gpio_init(int pin)
 }
 
 mraa_gpio_context
-mraa_gpio_chardev_init(int pins[], int num_pins)
+mraa_gpio_chardev_init(int pins[], int num_pins, unsigned int chardev_ver)
 {
     int chip_id, line_offset;
     mraa_gpio_context dev;
@@ -451,7 +442,9 @@ mraa_gpio_chardev_init(int pins[], int num_pins)
 
             gpio_group[chip_id].dev_fd = cinfo->chip_fd;
             gpio_group[chip_id].is_required = 1;
-            gpio_group[chip_id].gpiod_handle = -1;
+            if (chardev_ver == 1) {
+                gpio_group[chip_id].gpiod_handle = -1;
+            }
 
             free(cinfo);
         }
@@ -470,6 +463,23 @@ mraa_gpio_chardev_init(int pins[], int num_pins)
         gpio_group[chip_id].num_gpio_lines++;
     }
 
+    if (chardev_ver == 2) {
+        for (int i = 0; i < dev->num_chips; ++i) {
+            if (gpio_group[i].is_required) {
+                int line_request_fd =
+                mraa_gpiod_get_lines_request(gpio_group[i].dev_fd, gpio_group[i].gpio_lines,
+                                             gpio_group[i].num_gpio_lines);
+                if (line_request_fd < 0) {
+                    syslog(LOG_ERR, "[GPIOD_INTERFACE]: error requesting lines for chip %d", gpio_group[i].gpio_chip);
+                    mraa_gpio_close(dev);
+                    return NULL;
+                }
+
+                gpio_group[i].line_request_fd = line_request_fd;
+            }
+        }
+    }
+
     /* Initialize rw_values for read / write multiple functions.
      * Also, allocate memory for inverse map: */
     for (int i = 0; i < dev->num_chips; ++i) {
@@ -543,8 +553,8 @@ mraa_gpio_init_multi(int pins[], int num_pins)
         return NULL;
     }
 
-    if (board->chardev_capable)
-        return mraa_gpio_chardev_init(pins, num_pins);
+    if (board->chardev_ver)
+        return mraa_gpio_chardev_init(pins, num_pins, board->chardev_ver);
 
     /* Fallback to legacy interface. */
     mraa_gpio_context head = NULL, current, tmp;
@@ -652,13 +662,18 @@ mraa_gpio_wait_interrupt(int fds[],
 static mraa_result_t
 mraa_gpio_chardev_wait_interrupt(int fds[], int num_fds, mraa_gpio_events_t events)
 {
-    struct pollfd pfd[num_fds];
-    struct gpioevent_data event_data;
-
     if (!fds) {
         return MRAA_ERROR_INVALID_PARAMETER;
     }
 
+    struct gpio_v2_line_event event_v2;
+    struct gpioevent_data event_v1;
+    struct pollfd pfd[num_fds];
+    int ret;
+
+    memset(&event_v2, 0, sizeof(event_v2));
+    memset(&event_v1, 0, sizeof(event_v1));
+
     for (int i = 0; i < num_fds; ++i) {
         pfd[i].fd = fds[i];
         pfd[i].events = POLLIN;
@@ -670,11 +685,26 @@ mraa_gpio_chardev_wait_interrupt(int fds[], int num_fds, mraa_gpio_events_t even
 
     for (int i = 0; i < num_fds; ++i) {
         if (pfd[i].revents & POLLIN) {
-            read(fds[i], &event_data, sizeof(event_data));
+            void* event_buf;
+            size_t event_buf_size;
+            if (plat->chardev_ver == 2) {
+                event_buf = &event_v2;
+                event_buf_size = sizeof(event_v2);
+            } else {
+                event_buf = &event_v1;
+                event_buf_size = sizeof(event_v1);
+            }
+
+            ret = read(fds[i], event_buf, event_buf_size);
             events[i].id = i;
-            events[i].timestamp = event_data.timestamp;
-        } else
+            if (plat->chardev_ver == 2) {
+                events[i].timestamp = event_v2.timestamp_ns;
+            } else {
+                events[i].timestamp = event_v1.timestamp;
+            }
+        } else {
             events[i].id = -1;
+        }
     }
 
     return MRAA_SUCCESS;
@@ -690,7 +720,7 @@ mraa_gpio_get_events(mraa_gpio_context dev)
 
     unsigned int event_idx = 0;
 
-    if (plat->chardev_capable) {
+    if (plat->chardev_ver) {
         unsigned int pin_idx;
         mraa_gpiod_group_t gpio_iter;
 
@@ -746,14 +776,18 @@ mraa_gpio_interrupt_handler(void* arg)
     /* Is this pin on a subplatform? Do nothing... */
     if (mraa_is_sub_platform_id(dev->pin)) {
     }
-    /* Is the platform chardev_capable? */
-    else if (plat->chardev_capable) {
+    /* Is the platform chardev_ver? */
+    else if (plat->chardev_ver) {
         mraa_gpiod_group_t gpio_group;
 
         for_each_gpio_group(gpio_group, dev)
         {
-            for (int i = 0; i < gpio_group->num_gpio_lines; ++i) {
-                fps[idx++] = gpio_group->event_handles[i];
+            if (plat->chardev_ver == 2) {
+                fps[idx++] = gpio_group->line_request_fd;
+            } else {
+                for (int i = 0; i < gpio_group->num_gpio_lines; ++i) {
+                    fps[idx++] = gpio_group->event_handles[i];
+                }
             }
         }
     }
@@ -800,7 +834,7 @@ mraa_gpio_interrupt_handler(void* arg)
         if (IS_FUNC_DEFINED(dev, gpio_wait_interrupt_replace)) {
             ret = dev->advance_func->gpio_wait_interrupt_replace(dev);
         } else {
-            if (plat->chardev_capable) {
+            if (plat->chardev_ver) {
                 ret = mraa_gpio_chardev_wait_interrupt(fps, idx, dev->events);
             } else {
                 ret = mraa_gpio_wait_interrupt(fps, idx
@@ -852,7 +886,7 @@ mraa_gpio_chardev_edge_mode(mraa_gpio_context dev, mraa_gpio_edge_t mode)
         return MRAA_ERROR_INVALID_HANDLE;
     }
 
-    if (!plat->chardev_capable) {
+    if (!plat->chardev_ver) {
         syslog(LOG_ERR, "mraa_gpio_chardev_edge_mode() not supported for old sysfs interface");
         return MRAA_ERROR_FEATURE_NOT_IMPLEMENTED;
     }
@@ -860,48 +894,114 @@ mraa_gpio_chardev_edge_mode(mraa_gpio_context dev, mraa_gpio_edge_t mode)
     int status;
     mraa_gpiod_group_t gpio_group;
 
-    struct gpioevent_request req;
+    if (plat->chardev_ver == 2) {
+        for_each_gpio_group(gpio_group, dev)
+        {
+            unsigned long flags = 0;
 
-    switch (mode) {
-        case MRAA_GPIO_EDGE_BOTH:
-            req.eventflags = GPIOEVENT_REQUEST_BOTH_EDGES;
-            break;
-        case MRAA_GPIO_EDGE_RISING:
-            req.eventflags = GPIOEVENT_REQUEST_RISING_EDGE;
-            break;
-        case MRAA_GPIO_EDGE_FALLING:
-            req.eventflags = GPIOEVENT_REQUEST_FALLING_EDGE;
-            break;
-        /* Chardev interface doesn't handle EDGE_NONE. */
-        case MRAA_GPIO_EDGE_NONE:
-        default:
-            return MRAA_ERROR_FEATURE_NOT_IMPLEMENTED;
-    }
+            int ret = mraa_gpiod_get_line_flags(gpio_group->dev_fd, gpio_group->gpio_lines[0], &flags);
+            if (ret != 0) {
+                syslog(LOG_ERR, "[GPIOD_INTERFACE]: error getting line flags");
+                return MRAA_ERROR_INVALID_RESOURCE;
+            }
 
-    for_each_gpio_group(gpio_group, dev)
-    {
-        if (gpio_group->gpiod_handle != -1) {
-            close(gpio_group->gpiod_handle);
-            gpio_group->gpiod_handle = -1;
+            flags &= ~GPIO_V2_LINE_FLAG_USED;
+
+            if ((flags & GPIO_V2_LINE_FLAG_INPUT) == 0) {
+                syslog(LOG_ERR, "[GPIOD_INTERFACE]: error setting line input mode. Line is not "
+                                "configured as input.");
+                return MRAA_ERROR_INVALID_RESOURCE;
+            }
+
+            switch (mode) {
+                case MRAA_GPIO_EDGE_BOTH:
+                    if ((flags & GPIO_V2_LINE_FLAG_EDGE_RISING) && (flags & GPIO_V2_LINE_FLAG_EDGE_FALLING)) {
+                        syslog(LOG_DEBUG,
+                               "[GPIOD_INTERFACE]: line already configured for both edges");
+                        return MRAA_SUCCESS;
+                    }
+                    flags |= (GPIO_V2_LINE_FLAG_EDGE_RISING | GPIO_V2_LINE_FLAG_EDGE_FALLING);
+                    break;
+                case MRAA_GPIO_EDGE_RISING:
+                    if ((flags & GPIO_V2_LINE_FLAG_EDGE_RISING) && !(flags & GPIO_V2_LINE_FLAG_EDGE_FALLING)) {
+                        syslog(LOG_DEBUG,
+                               "[GPIOD_INTERFACE]: line already configured for rising edge");
+                        return MRAA_SUCCESS;
+                    }
+                    flags |= GPIO_V2_LINE_FLAG_EDGE_RISING;
+                    flags &= ~GPIO_V2_LINE_FLAG_EDGE_FALLING;
+                    break;
+                case MRAA_GPIO_EDGE_FALLING:
+                    if ((flags & GPIO_V2_LINE_FLAG_EDGE_FALLING) && !(flags & GPIO_V2_LINE_FLAG_EDGE_RISING)) {
+                        syslog(LOG_DEBUG,
+                               "[GPIOD_INTERFACE]: line already configured for falling edge");
+                        return MRAA_SUCCESS;
+                    }
+                    flags |= GPIO_V2_LINE_FLAG_EDGE_FALLING;
+                    flags &= ~GPIO_V2_LINE_FLAG_EDGE_RISING;
+                    break;
+                case MRAA_GPIO_EDGE_NONE:
+                    if (!(flags & GPIO_V2_LINE_FLAG_EDGE_RISING) && !(flags & GPIO_V2_LINE_FLAG_EDGE_FALLING)) {
+                        syslog(LOG_DEBUG,
+                               "[GPIOD_INTERFACE]: line already configured for no edge detection");
+                        return MRAA_SUCCESS;
+                    }
+                    flags &= ~(GPIO_V2_LINE_FLAG_EDGE_RISING | GPIO_V2_LINE_FLAG_EDGE_FALLING);
+                    break;
+                default:
+                    return MRAA_ERROR_FEATURE_NOT_IMPLEMENTED;
+            }
+
+            ret = mraa_gpiod_reconfig_lines(gpio_group->line_request_fd, flags);
+            if (ret != 0) {
+                syslog(LOG_ERR, "[GPIOD_INTERFACE]: error reconfiguring line flags: %d", ret);
+                return MRAA_ERROR_INVALID_RESOURCE;
+            }
         }
+    } else {
+        struct gpioevent_request req;
 
-        gpio_group->event_handles = malloc(gpio_group->num_gpio_lines * sizeof(int));
-        if (!gpio_group->event_handles) {
-            syslog(LOG_ERR, "mraa_gpio_chardev_edge_mode(): malloc error!");
-            return MRAA_ERROR_NO_RESOURCES;
+        switch (mode) {
+            case MRAA_GPIO_EDGE_BOTH:
+                req.eventflags = GPIOEVENT_REQUEST_BOTH_EDGES;
+                break;
+            case MRAA_GPIO_EDGE_RISING:
+                req.eventflags = GPIOEVENT_REQUEST_RISING_EDGE;
+                break;
+            case MRAA_GPIO_EDGE_FALLING:
+                req.eventflags = GPIOEVENT_REQUEST_FALLING_EDGE;
+                break;
+            /* Chardev interface doesn't handle EDGE_NONE. */
+            case MRAA_GPIO_EDGE_NONE:
+            default:
+                return MRAA_ERROR_FEATURE_NOT_IMPLEMENTED;
         }
 
-        for (int i = 0; i < gpio_group->num_gpio_lines; ++i) {
-            req.lineoffset = gpio_group->gpio_lines[i];
-            req.handleflags = GPIOHANDLE_REQUEST_INPUT;
+        for_each_gpio_group(gpio_group, dev)
+        {
+            if (gpio_group->gpiod_handle != -1) {
+                close(gpio_group->gpiod_handle);
+                gpio_group->gpiod_handle = -1;
+            }
 
-            status = _mraa_gpiod_ioctl(gpio_group->dev_fd, GPIO_GET_LINEEVENT_IOCTL, &req);
-            if (status < 0) {
-                syslog(LOG_ERR, "error getting line event handle for line %i", gpio_group->gpio_lines[i]);
-                return MRAA_ERROR_INVALID_RESOURCE;
+            gpio_group->event_handles = malloc(gpio_group->num_gpio_lines * sizeof(int));
+            if (!gpio_group->event_handles) {
+                syslog(LOG_ERR, "mraa_gpio_chardev_edge_mode(): malloc error!");
+                return MRAA_ERROR_NO_RESOURCES;
             }
 
-            gpio_group->event_handles[i] = req.fd;
+            for (int i = 0; i < gpio_group->num_gpio_lines; ++i) {
+                req.lineoffset = gpio_group->gpio_lines[i];
+                req.handleflags = GPIOHANDLE_REQUEST_INPUT;
+
+                status = _mraa_gpiod_ioctl(gpio_group->dev_fd, GPIO_GET_LINEEVENT_IOCTL, &req);
+                if (status < 0) {
+                    syslog(LOG_ERR, "error getting line event handle for line %i", gpio_group->gpio_lines[i]);
+                    return MRAA_ERROR_INVALID_RESOURCE;
+                }
+
+                gpio_group->event_handles[i] = req.fd;
+            }
         }
     }
 
@@ -932,7 +1032,7 @@ mraa_gpio_edge_mode(mraa_gpio_context dev, mraa_gpio_edge_t mode)
         }
     }
 
-    if (plat->chardev_capable)
+    if (plat->chardev_ver)
         return mraa_gpio_chardev_edge_mode(dev, mode);
 
     mraa_gpio_context it = dev;
@@ -1051,7 +1151,7 @@ mraa_gpio_isr_exit(mraa_gpio_context dev)
     dev->isr_thread_terminating = 1;
 
     // stop isr being useful
-    if (plat && (plat->chardev_capable))
+    if (plat && (plat->chardev_ver))
         _mraa_close_gpio_event_handles(dev);
     else
         ret = mraa_gpio_edge_mode(dev, MRAA_GPIO_EDGE_NONE);
@@ -1101,8 +1201,8 @@ mraa_gpio_mode(mraa_gpio_context dev, mraa_gpio_mode_t mode)
             return pre_ret;
     }
 
-    if (plat->chardev_capable) {
-        unsigned flags = 0;
+    if (plat->chardev_ver) {
+        unsigned long flags = 0;
         int line_handle;
         mraa_gpiod_group_t gpio_iter;
 
@@ -1111,44 +1211,87 @@ mraa_gpio_mode(mraa_gpio_context dev, mraa_gpio_mode_t mode)
         /* We save flag values from the first valid line. */
         for_each_gpio_group(gpio_iter, dev)
         {
-            mraa_gpiod_line_info* linfo =
-            mraa_get_line_info_by_chip_number(gpio_iter->gpio_chip, gpio_iter->gpio_lines[0]);
-            if (!linfo) {
-                syslog(LOG_ERR, "[GPIOD_INTERFACE]: error getting line info");
-                return MRAA_ERROR_UNSPECIFIED;
+            if (plat->chardev_ver == 2) {
+                int ret = mraa_gpiod_get_line_flags(gpio_iter->dev_fd, gpio_iter->gpio_lines[0], &flags);
+                if (ret != 0) {
+                    syslog(LOG_ERR, "[GPIOD_INTERFACE]: error getting line flags");
+                    return MRAA_ERROR_INVALID_RESOURCE;
+                }
+                flags &= ~GPIO_V2_LINE_FLAG_USED;
+            } else {
+                mraa_gpiod_line_info* linfo =
+                mraa_get_line_info_by_chip_number(gpio_iter->gpio_chip, gpio_iter->gpio_lines[0]);
+                if (!linfo) {
+                    syslog(LOG_ERR, "[GPIOD_INTERFACE]: error getting line info");
+                    return MRAA_ERROR_UNSPECIFIED;
+                }
+                flags = linfo->flags;
+                free(linfo);
             }
-            flags = linfo->flags;
-            free(linfo);
 
             /* We don't need to iterate any further. */
             break;
         }
 
         /* Without changing the API, for now, we can request only one mode per call. */
-        switch (mode) {
-            case MRAA_GPIOD_ACTIVE_LOW:
-                flags |= GPIOHANDLE_REQUEST_ACTIVE_LOW;
-                break;
-            case MRAA_GPIOD_OPEN_DRAIN:
-                flags |= GPIOHANDLE_REQUEST_OPEN_DRAIN;
-                break;
-            case MRAA_GPIOD_OPEN_SOURCE:
-                flags |= GPIOHANDLE_REQUEST_OPEN_SOURCE;
-                break;
-            default:
-                return MRAA_ERROR_FEATURE_NOT_IMPLEMENTED;
+        if (plat->chardev_ver == 2) {
+            switch (mode) {
+                case MRAA_GPIO_PULLUP:
+                    flags |= GPIO_V2_LINE_FLAG_BIAS_PULL_UP;
+                    break;
+                case MRAA_GPIO_PULLDOWN:
+                    flags |= GPIO_V2_LINE_FLAG_BIAS_PULL_DOWN;
+                    break;
+                case MRAA_GPIO_STRONG:
+                case MRAA_GPIO_HIZ:
+                    flags |= GPIO_V2_LINE_FLAG_BIAS_DISABLED;
+                    break;
+                case MRAA_GPIOD_ACTIVE_LOW:
+                    flags |= GPIO_V2_LINE_FLAG_ACTIVE_LOW;
+                    break;
+                case MRAA_GPIOD_OPEN_DRAIN:
+                    flags |= GPIO_V2_LINE_FLAG_OPEN_DRAIN;
+                    break;
+                case MRAA_GPIOD_OPEN_SOURCE:
+                    flags |= GPIO_V2_LINE_FLAG_OPEN_SOURCE;
+                    break;
+                default:
+                    return MRAA_ERROR_FEATURE_NOT_IMPLEMENTED;
+            }
+        } else {
+            switch (mode) {
+                case MRAA_GPIOD_ACTIVE_LOW:
+                    flags |= GPIOHANDLE_REQUEST_ACTIVE_LOW;
+                    break;
+                case MRAA_GPIOD_OPEN_DRAIN:
+                    flags |= GPIOHANDLE_REQUEST_OPEN_DRAIN;
+                    break;
+                case MRAA_GPIOD_OPEN_SOURCE:
+                    flags |= GPIOHANDLE_REQUEST_OPEN_SOURCE;
+                    break;
+                default:
+                    return MRAA_ERROR_FEATURE_NOT_IMPLEMENTED;
+            }
         }
 
         for_each_gpio_group(gpio_iter, dev)
         {
-            line_handle = mraa_get_lines_handle(gpio_iter->dev_fd, gpio_iter->gpio_lines,
-                                                gpio_iter->num_gpio_lines, flags, 0);
-            if (line_handle <= 0) {
-                syslog(LOG_ERR, "[GPIOD_INTERFACE]: error getting line handle");
-                return MRAA_ERROR_INVALID_RESOURCE;
-            }
+            if (plat->chardev_ver == 2) {
+                int ret = mraa_gpiod_reconfig_lines(gpio_iter->line_request_fd, flags);
+                if (ret != 0) {
+                    syslog(LOG_ERR, "[GPIOD_INTERFACE]: error reconfiguring line flags: %d", ret);
+                    return MRAA_ERROR_INVALID_RESOURCE;
+                }
+            } else {
+                line_handle = mraa_get_lines_handle(gpio_iter->dev_fd, gpio_iter->gpio_lines,
+                                                    gpio_iter->num_gpio_lines, flags, 0);
+                if (line_handle <= 0) {
+                    syslog(LOG_ERR, "[GPIOD_INTERFACE]: error getting line handle");
+                    return MRAA_ERROR_INVALID_RESOURCE;
+                }
 
-            gpio_iter->gpiod_handle = line_handle;
+                gpio_iter->gpiod_handle = line_handle;
+            }
         }
     } else {
 
@@ -1200,8 +1343,8 @@ mraa_gpio_mode(mraa_gpio_context dev, mraa_gpio_mode_t mode)
     return MRAA_SUCCESS;
 }
 
-mraa_result_t
-mraa_gpio_chardev_dir(mraa_gpio_context dev, mraa_gpio_dir_t dir)
+static mraa_result_t
+mraa_gpio_chardev_dir_v1(mraa_gpio_context dev, mraa_gpio_dir_t dir)
 {
     int line_handle;
     unsigned flags = 0;
@@ -1266,6 +1409,81 @@ mraa_gpio_chardev_dir(mraa_gpio_context dev, mraa_gpio_dir_t dir)
     return MRAA_SUCCESS;
 }
 
+static mraa_result_t
+mraa_gpio_chardev_dir_v2(mraa_gpio_context dev, mraa_gpio_dir_t dir)
+{
+    unsigned long flags = 0;
+    struct _gpio_group* gpio_iter;
+
+    for_each_gpio_group(gpio_iter, dev)
+    {
+        int ret = mraa_gpiod_get_line_flags(gpio_iter->dev_fd, gpio_iter->gpio_lines[0], &flags);
+        if (ret != 0) {
+            syslog(LOG_ERR, "[GPIOD_INTERFACE]: error getting line flags");
+            return MRAA_ERROR_INVALID_RESOURCE;
+        }
+
+        int need_reconfig = 1;
+
+        switch (dir) {
+            case MRAA_GPIO_OUT_HIGH:
+            case MRAA_GPIO_OUT_LOW:
+            case MRAA_GPIO_OUT:
+                if (flags & GPIO_V2_LINE_FLAG_OUTPUT) {
+                    syslog(LOG_DEBUG, "[GPIOD_INTERFACE]: line already configured as output");
+                    need_reconfig = 0;
+                    break;
+                }
+                flags |= GPIO_V2_LINE_FLAG_OUTPUT;
+                flags &= ~GPIO_V2_LINE_FLAG_INPUT;
+                break;
+            case MRAA_GPIO_IN:
+                if (flags & GPIO_V2_LINE_FLAG_INPUT) {
+                    syslog(LOG_DEBUG, "[GPIOD_INTERFACE]: line already configured as input");
+                    need_reconfig = 0;
+                    break;
+                }
+                flags |= GPIO_V2_LINE_FLAG_INPUT;
+                flags &= ~GPIO_V2_LINE_FLAG_OUTPUT;
+                break;
+            default:
+                return MRAA_ERROR_FEATURE_NOT_IMPLEMENTED;
+        }
+
+        if (need_reconfig) {
+            flags &= ~GPIO_V2_LINE_FLAG_USED;
+            ret = mraa_gpiod_reconfig_lines(gpio_iter->line_request_fd, flags);
+            if (ret != 0) {
+                syslog(LOG_ERR, "[GPIOD_INTERFACE]: error reconfiguring line flags: %d", ret);
+                return MRAA_ERROR_INVALID_RESOURCE;
+            }
+        }
+
+        if ((dir == MRAA_GPIO_OUT_LOW) || (dir == MRAA_GPIO_OUT_HIGH)) {
+            unsigned char* values = (unsigned char*) calloc(gpio_iter->num_gpio_lines, sizeof(unsigned char));
+            if (!values) {
+                syslog(LOG_ERR, "[GPIOD_INTERFACE]: error allocating memory for values");
+                return MRAA_ERROR_NO_RESOURCES;
+            }
+
+            if (dir == MRAA_GPIO_OUT_LOW) {
+                memset(values, 0, gpio_iter->num_gpio_lines);
+            } else {
+                memset(values, 1, gpio_iter->num_gpio_lines);
+            }
+
+            ret = mraa_gpiod_set_values(gpio_iter->line_request_fd, gpio_iter->num_gpio_lines, values);
+            free(values);
+            if (ret != 0) {
+                syslog(LOG_ERR, "[GPIOD_INTERFACE]: error setting line values: %d", ret);
+                return MRAA_ERROR_INVALID_RESOURCE;
+            }
+        }
+    }
+
+    return MRAA_SUCCESS;
+}
+
 static mraa_result_t
 gpio_sysfs_read_dir(mraa_gpio_context dev, int dir_fd, mraa_gpio_dir_t *dir)
 {
@@ -1310,8 +1528,9 @@ mraa_gpio_dir(mraa_gpio_context dev, mraa_gpio_dir_t dir)
         }
     }
 
-    if (plat->chardev_capable)
-        return mraa_gpio_chardev_dir(dev, dir);
+    if (plat->chardev_ver)
+        return plat->chardev_ver == 2 ? mraa_gpio_chardev_dir_v2(dev, dir) :
+                                            mraa_gpio_chardev_dir_v1(dev, dir);
 
     mraa_gpio_context it = dev;
 
@@ -1390,36 +1609,51 @@ mraa_gpio_read_dir(mraa_gpio_context dev, mraa_gpio_dir_t* dir)
     mraa_result_t result = MRAA_SUCCESS;
 
     /* Initialize with 'unusable'. */
-    unsigned flags = GPIOLINE_FLAG_KERNEL;
+    unsigned long flags = GPIOLINE_FLAG_KERNEL;
+
+    if (plat->chardev_ver == 2)
+        flags = 0;
 
     if (IS_FUNC_DEFINED(dev, gpio_read_dir_replace)) {
         return dev->advance_func->gpio_read_dir_replace(dev, dir);
     }
 
-    if (plat->chardev_capable) {
+    if (plat->chardev_ver) {
         mraa_gpiod_group_t gpio_iter;
 
         for_each_gpio_group(gpio_iter, dev)
         {
-            mraa_gpiod_line_info* linfo =
-            mraa_get_line_info_by_chip_number(gpio_iter->gpio_chip, gpio_iter->gpio_lines[0]);
-            if (!linfo) {
-                syslog(LOG_ERR, "[GPIOD_INTERFACE]: error getting line info");
-                return MRAA_ERROR_UNSPECIFIED;
+            if (plat->chardev_ver == 2) {
+                int ret = mraa_gpiod_get_line_flags(gpio_iter->dev_fd, gpio_iter->gpio_lines[0], &flags);
+                if (ret != 0) {
+                    syslog(LOG_ERR, "[GPIOD_INTERFACE]: error getting line flags");
+                    return MRAA_ERROR_INVALID_RESOURCE;
+                }
+            } else {
+                mraa_gpiod_line_info* linfo =
+                mraa_get_line_info_by_chip_number(gpio_iter->gpio_chip, gpio_iter->gpio_lines[0]);
+                if (!linfo) {
+                    syslog(LOG_ERR, "[GPIOD_INTERFACE]: error getting line info");
+                    return MRAA_ERROR_UNSPECIFIED;
+                }
+                flags = linfo->flags;
+                free(linfo);
             }
-            flags = linfo->flags;
-            free(linfo);
 
             /* We don't need to iterate further. */
             break;
         }
 
-        if (flags & GPIOLINE_FLAG_KERNEL) {
+        if ((plat->chardev_ver != 2) && (flags & GPIOLINE_FLAG_KERNEL)) {
             syslog(LOG_ERR, "[GPIOD_INTERFACE]: cannot read gpio direction. Line used by kernel.");
             return MRAA_ERROR_UNSPECIFIED;
         }
 
-        *dir = flags & GPIOLINE_FLAG_IS_OUT ? MRAA_GPIO_OUT : MRAA_GPIO_IN;
+        if (plat->chardev_ver == 2) {
+            *dir = flags & GPIO_V2_LINE_FLAG_OUTPUT ? MRAA_GPIO_OUT : MRAA_GPIO_IN;
+        } else {
+            *dir = flags & GPIOLINE_FLAG_IS_OUT ? MRAA_GPIO_OUT : MRAA_GPIO_IN;
+        }
     } else {
         char filepath[MAX_SIZE];
         int fd;
@@ -1461,7 +1695,7 @@ mraa_gpio_read(mraa_gpio_context dev)
         return dev->advance_func->gpio_read_replace(dev);
     }
 
-    if (plat->chardev_capable) {
+    if (plat->chardev_ver) {
         int output_values[1] = { 0 };
 
         if (mraa_gpio_read_multi(dev, output_values) != MRAA_SUCCESS)
@@ -1501,7 +1735,7 @@ mraa_gpio_read_multi(mraa_gpio_context dev, int output_values[])
         return -1;
     }
 
-    if (plat->chardev_capable) {
+    if (plat->chardev_ver) {
         memset(output_values, 0, dev->num_pins * sizeof(int));
 
         mraa_gpiod_group_t gpio_iter;
@@ -1509,22 +1743,32 @@ mraa_gpio_read_multi(mraa_gpio_context dev, int output_values[])
         for_each_gpio_group(gpio_iter, dev)
         {
             int status;
-            unsigned flags = GPIOHANDLE_REQUEST_INPUT;
 
-            if (gpio_iter->gpiod_handle <= 0) {
-                gpio_iter->gpiod_handle = mraa_get_lines_handle(gpio_iter->dev_fd, gpio_iter->gpio_lines,
-                                                                gpio_iter->num_gpio_lines, flags, 0);
+            if (plat->chardev_ver == 2) {
+                status = mraa_gpiod_get_values(gpio_iter->line_request_fd,
+                                               gpio_iter->num_gpio_lines, gpio_iter->rw_values);
+                if (status < 0) {
+                    syslog(LOG_ERR, "[GPIOD_INTERFACE]: error reading gpio");
+                    return MRAA_ERROR_INVALID_RESOURCE;
+                }
+            } else {
+                unsigned flags = GPIOHANDLE_REQUEST_INPUT;
+
                 if (gpio_iter->gpiod_handle <= 0) {
-                    syslog(LOG_ERR, "[GPIOD_INTERFACE]: error getting gpio line handle");
-                    return MRAA_ERROR_INVALID_HANDLE;
+                    gpio_iter->gpiod_handle = mraa_get_lines_handle(gpio_iter->dev_fd, gpio_iter->gpio_lines,
+                                                                    gpio_iter->num_gpio_lines, flags, 0);
+                    if (gpio_iter->gpiod_handle <= 0) {
+                        syslog(LOG_ERR, "[GPIOD_INTERFACE]: error getting gpio line handle");
+                        return MRAA_ERROR_INVALID_HANDLE;
+                    }
                 }
-            }
 
-            status =
-            mraa_get_line_values(gpio_iter->gpiod_handle, gpio_iter->num_gpio_lines, gpio_iter->rw_values);
-            if (status < 0) {
-                syslog(LOG_ERR, "[GPIOD_INTERFACE]: error writing gpio");
-                return MRAA_ERROR_INVALID_RESOURCE;
+                status = mraa_get_line_values(gpio_iter->gpiod_handle, gpio_iter->num_gpio_lines,
+                                              gpio_iter->rw_values);
+                if (status < 0) {
+                    syslog(LOG_ERR, "[GPIOD_INTERFACE]: error writing gpio");
+                    return MRAA_ERROR_INVALID_RESOURCE;
+                }
             }
 
             /* Write values back to the user provided array. */
@@ -1571,7 +1815,7 @@ mraa_gpio_write(mraa_gpio_context dev, int value)
         return dev->advance_func->gpio_write_replace(dev, value);
     }
 
-    if (plat->chardev_capable) {
+    if (plat->chardev_ver) {
         int input_values[1] = { value };
 
         return mraa_gpio_write_multi(dev, input_values);
@@ -1614,7 +1858,7 @@ mraa_gpio_write_multi(mraa_gpio_context dev, int input_values[])
         return MRAA_ERROR_INVALID_HANDLE;
     }
 
-    if (plat->chardev_capable) {
+    if (plat->chardev_ver) {
         mraa_gpiod_group_t gpio_iter;
 
         int* counters = calloc(dev->num_chips, sizeof(int));
@@ -1634,22 +1878,31 @@ mraa_gpio_write_multi(mraa_gpio_context dev, int input_values[])
         for_each_gpio_group(gpio_iter, dev)
         {
             int status;
-            unsigned flags = GPIOHANDLE_REQUEST_OUTPUT;
+            if (plat->chardev_ver == 2) {
+                status = mraa_gpiod_set_values(gpio_iter->line_request_fd,
+                                               gpio_iter->num_gpio_lines, gpio_iter->rw_values);
+                if (status < 0) {
+                    syslog(LOG_ERR, "[GPIOD_INTERFACE]: error writing gpio");
+                    return MRAA_ERROR_INVALID_RESOURCE;
+                }
+            } else {
+                unsigned flags = GPIOHANDLE_REQUEST_OUTPUT;
 
-            if (gpio_iter->gpiod_handle <= 0) {
-                gpio_iter->gpiod_handle = mraa_get_lines_handle(gpio_iter->dev_fd, gpio_iter->gpio_lines,
-                                                                gpio_iter->num_gpio_lines, flags, 0);
                 if (gpio_iter->gpiod_handle <= 0) {
-                    syslog(LOG_ERR, "[GPIOD_INTERFACE]: error getting gpio line handle");
-                    return MRAA_ERROR_INVALID_HANDLE;
+                    gpio_iter->gpiod_handle = mraa_get_lines_handle(gpio_iter->dev_fd, gpio_iter->gpio_lines,
+                                                                    gpio_iter->num_gpio_lines, flags, 0);
+                    if (gpio_iter->gpiod_handle <= 0) {
+                        syslog(LOG_ERR, "[GPIOD_INTERFACE]: error getting gpio line handle");
+                        return MRAA_ERROR_INVALID_HANDLE;
+                    }
                 }
-            }
 
-            status =
-            mraa_set_line_values(gpio_iter->gpiod_handle, gpio_iter->num_gpio_lines, gpio_iter->rw_values);
-            if (status < 0) {
-                syslog(LOG_ERR, "[GPIOD_INTERFACE]: error writing gpio");
-                return MRAA_ERROR_INVALID_RESOURCE;
+                status = mraa_set_line_values(gpio_iter->gpiod_handle, gpio_iter->num_gpio_lines,
+                                              gpio_iter->rw_values);
+                if (status < 0) {
+                    syslog(LOG_ERR, "[GPIOD_INTERFACE]: error writing gpio");
+                    return MRAA_ERROR_INVALID_RESOURCE;
+                }
             }
         }
     } else {
@@ -1754,7 +2007,7 @@ mraa_gpio_close(mraa_gpio_context dev)
     /* Free any ISRs */
     mraa_gpio_isr_exit(dev);
 
-    if (plat && plat->chardev_capable) {
+    if (plat && plat->chardev_ver) {
         _mraa_free_gpio_groups(dev);
 
         free(dev);
@@ -1832,6 +2085,60 @@ mraa_gpio_get_pin_raw(mraa_gpio_context dev)
     return dev->pin;
 }
 
+mraa_result_t
+mraa_gpio_input_mode_chardev(mraa_gpio_context dev, mraa_gpio_input_mode_t mode)
+{
+    mraa_gpiod_group_t gpio_group;
+    unsigned long flags = 0;
+    int ret;
+
+    for_each_gpio_group(gpio_group, dev)
+    {
+        ret = mraa_gpiod_get_line_flags(gpio_group->dev_fd, gpio_group->gpio_lines[0], &flags);
+        if (ret != 0) {
+            syslog(LOG_ERR, "[GPIOD_INTERFACE]: error getting line flags");
+            return MRAA_ERROR_INVALID_RESOURCE;
+        }
+
+        if ((flags & GPIO_V2_LINE_FLAG_INPUT) == 0) {
+            syslog(LOG_ERR, "[GPIOD_INTERFACE]: error setting line input mode. Line is not "
+                            "configured as input.");
+            return MRAA_ERROR_INVALID_RESOURCE;
+        }
+
+        switch (mode) {
+            case MRAA_GPIO_ACTIVE_HIGH:
+                if ((flags & GPIO_V2_LINE_FLAG_ACTIVE_LOW) == 0) {
+                    syslog(LOG_DEBUG, "[GPIOD_INTERFACE]: Line is already active high.");
+                    return MRAA_SUCCESS;
+                }
+                flags &= ~GPIO_V2_LINE_FLAG_ACTIVE_LOW;
+                break;
+            case MRAA_GPIO_ACTIVE_LOW:
+                if (flags & GPIO_V2_LINE_FLAG_ACTIVE_LOW) {
+                    syslog(LOG_DEBUG, "[GPIOD_INTERFACE]: Line is already active low.");
+                    return MRAA_SUCCESS;
+                }
+                flags |= GPIO_V2_LINE_FLAG_ACTIVE_LOW;
+                break;
+            default:
+                return MRAA_ERROR_FEATURE_NOT_IMPLEMENTED;
+        }
+
+        flags &= ~GPIO_V2_LINE_FLAG_USED;
+
+        ret = mraa_gpiod_reconfig_lines(gpio_group->line_request_fd, flags);
+        if (ret != 0) {
+            syslog(LOG_ERR, "[GPIOD_INTERFACE]: error reconfiguring line flags: %d", ret);
+            return MRAA_ERROR_INVALID_RESOURCE;
+        }
+
+        break;
+    }
+
+    return MRAA_SUCCESS;
+}
+
 mraa_result_t
 mraa_gpio_input_mode(mraa_gpio_context dev, mraa_gpio_input_mode_t mode)
 {
@@ -1840,6 +2147,10 @@ mraa_gpio_input_mode(mraa_gpio_context dev, mraa_gpio_input_mode_t mode)
         return MRAA_ERROR_INVALID_HANDLE;
     }
 
+    if (plat->chardev_ver == 2) {
+        return mraa_gpio_input_mode_chardev(dev, mode);
+    }
+
     char filepath[MAX_SIZE];
     snprintf(filepath, MAX_SIZE, SYSFS_CLASS_GPIO "/gpio%d/active_low", dev->pin);
 
diff --git a/src/gpio/gpio_chardev.c b/src/gpio/gpio_chardev.c
index 0c4636ea0701..a3e8882eddb4 100644
--- a/src/gpio/gpio_chardev.c
+++ b/src/gpio/gpio_chardev.c
@@ -109,6 +109,10 @@ _mraa_close_gpio_desc(mraa_gpio_context dev)
 {
     mraa_gpiod_group_t gpio_iter;
 
+    if (plat->chardev_ver == 2) {
+        return;
+    }
+
     for_each_gpio_group(gpio_iter, dev)
     {
         if (gpio_iter->gpiod_handle != -1) {
diff --git a/src/gpio/gpio_chardev_v2.c b/src/gpio/gpio_chardev_v2.c
new file mode 100644
index 000000000000..d1d4defbb894
--- /dev/null
+++ b/src/gpio/gpio_chardev_v2.c
@@ -0,0 +1,185 @@
+/*
+ * Author: Baocheng Su <baocheng.su@siemens.com>
+ * Copyright (c) 2026 Siemens AG
+ *
+ * SPDX-License-Identifier: MIT
+ */
+
+#include "gpio/gpio_chardev_v2.h"
+#include "linux/gpio.h"
+#include "mraa_internal.h"
+
+#include <string.h>
+
+int
+mraa_gpiod_get_lines_request(int chip_fd, unsigned int line_offsets[], unsigned num_lines)
+{
+    struct gpio_v2_line_config config;
+    struct gpio_v2_line_request req;
+
+    memset(&config, 0, sizeof(config));
+    memset(&req, 0, sizeof(req));
+    req.config = config;
+    strcpy(req.consumer, "mraa");
+    req.num_lines = num_lines;
+    memcpy(req.offsets, line_offsets, num_lines * sizeof(unsigned int));
+
+    int status = _mraa_gpiod_ioctl(chip_fd, GPIO_V2_GET_LINE_IOCTL, &req);
+    if (status < 0) {
+        syslog(LOG_ERR, "gpiod: ioctl() GPIO_V2_GET_LINE_IOCTL fail");
+        return status;
+    }
+
+    return req.fd;
+}
+
+int
+mraa_gpiod_get_line_name(int chip_fd, unsigned line_offset, char* name, size_t name_size)
+{
+    struct gpio_v2_line_info info;
+
+    memset(&info, 0, sizeof(info));
+    info.offset = line_offset;
+
+    int status = _mraa_gpiod_ioctl(chip_fd, GPIO_V2_GET_LINEINFO_IOCTL, &info);
+    if (status < 0) {
+        syslog(LOG_ERR, "[GPIOD_INTERFACE]: ioctl() GPIO_V2_GET_LINEINFO_IOCTL fail");
+        return status;
+    }
+
+    strncpy(name, info.name, name_size);
+    name[name_size - 1] = '\0';
+
+    return status;
+}
+
+int
+mraa_gpiod_get_line_flags(int chip_fd, unsigned line_offset, unsigned long* flag)
+{
+    struct gpio_v2_line_info info;
+
+    memset(&info, 0, sizeof(info));
+    info.offset = line_offset;
+
+    int status = _mraa_gpiod_ioctl(chip_fd, GPIO_V2_GET_LINEINFO_IOCTL, &info);
+    if (status < 0) {
+        syslog(LOG_ERR, "[GPIOD_INTERFACE]: ioctl() GPIO_V2_GET_LINEINFO_IOCTL fail");
+    }
+
+    *flag = info.flags;
+
+    return status;
+}
+
+
+int
+mraa_gpiod_find_gpio_line_by_name(const char* name, unsigned* chip_number, unsigned* line_number)
+{
+    char line_name[GPIO_MAX_NAME_SIZE];
+    mraa_gpiod_chip_info** cinfos;
+    mraa_gpiod_chip_info* cinfo;
+    int ret = -1;
+
+    int num_chips = mraa_get_chip_infos(&cinfos);
+    if (num_chips < 0) {
+        return -1;
+    }
+
+    for_each_gpio_chip(cinfo, cinfos, num_chips)
+    {
+        for (int i = 0; i < cinfo->chip_info.lines; i++) {
+            ret = mraa_gpiod_get_line_name(cinfo->chip_fd, i, line_name, GPIO_MAX_NAME_SIZE);
+            if (ret < 0) {
+                syslog(LOG_ERR, "[GPIOD_INTERFACE]: error getting line name for chip %d line %d", idx, i);
+                continue;
+            }
+
+            if (!strncmp(line_name, name, GPIO_MAX_NAME_SIZE)) {
+                if (chip_number) {
+                    /* idx is coming from `for_each_gpio_chip` definition */
+                    *chip_number = idx;
+                }
+
+                if (line_number) {
+                    *line_number = i;
+                }
+
+                ret = 0;
+                goto out;
+            }
+
+        }
+    }
+
+out:
+    for_each_gpio_chip(cinfo, cinfos, num_chips)
+    {
+        if (cinfo)
+            close(cinfo->chip_fd);
+    }
+    free(cinfos);
+    return ret;
+}
+
+int
+mraa_gpiod_get_values(int line_req_fd, unsigned num_lines, unsigned char line_values[])
+{
+    struct gpio_v2_line_values values;
+
+    memset(&values, 0, sizeof(values));
+
+    for (unsigned i = 0; i < num_lines; i++) {
+        values.mask |= (1ULL << i);
+    }
+
+    int status = _mraa_gpiod_ioctl(line_req_fd, GPIO_V2_LINE_GET_VALUES_IOCTL, &values);
+    if (status < 0) {
+        syslog(LOG_ERR, "[GPIOD_INTERFACE]: ioctl() GPIO_V2_LINE_GET_VALUES_IOCTL fail");
+        return status;
+    }
+
+    for (unsigned i = 0; i < num_lines; i++) {
+        line_values[i] = (values.bits & (1ULL << i)) ? 1 : 0;
+    }
+
+    return status;
+}
+
+int
+mraa_gpiod_set_values(int line_req_fd, unsigned num_lines, unsigned char line_values[])
+{
+    struct gpio_v2_line_values values;
+
+    memset(&values, 0, sizeof(values));
+    for (unsigned i = 0; i < num_lines; i++) {
+        if (line_values[i]) {
+            values.bits |= (1ULL << i);
+        }
+        values.mask |= (1ULL << i);
+    }
+
+    int status = _mraa_gpiod_ioctl(line_req_fd, GPIO_V2_LINE_SET_VALUES_IOCTL, &values);
+    if (status < 0) {
+        syslog(LOG_ERR, "[GPIOD_INTERFACE]: ioctl() GPIO_V2_LINE_SET_VALUES_IOCTL fail");
+        return status;
+    }
+
+    return status;
+}
+
+int
+mraa_gpiod_reconfig_lines(int line_req_fd, unsigned long flag)
+{
+    struct gpio_v2_line_config config;
+
+    memset(&config, 0, sizeof(config));
+    config.flags = flag;
+
+    int status = _mraa_gpiod_ioctl(line_req_fd, GPIO_V2_LINE_SET_CONFIG_IOCTL, &config);
+    if (status < 0) {
+        syslog(LOG_ERR, "[GPIOD_INTERFACE]: ioctl() GPIO_V2_LINE_SET_CONFIG_IOCTL fail");
+        return status;
+    }
+
+    return status;
+}
diff --git a/src/mraa.c b/src/mraa.c
index 927d9ab4af9b..fb7ee96d2452 100644
--- a/src/mraa.c
+++ b/src/mraa.c
@@ -95,11 +95,11 @@ mraa_is_kernel_chardev_interface_compatible()
     return 1;
 }
 
-mraa_boolean_t
-mraa_is_platform_chardev_interface_capable()
+static mraa_boolean_t
+mraa_check_chardev_ver()
 {
-    if ((plat != NULL) && (plat->chardev_capable)) {
-        return mraa_is_kernel_chardev_interface_compatible();
+    if ((plat != NULL) && (plat->chardev_ver)) {
+        return (mraa_is_kernel_chardev_interface_compatible()) ? plat->chardev_ver : 0;
     }
 
     syslog(LOG_NOTICE, "gpio: platform doesn't support chardev, falling back to sysfs");
@@ -229,8 +229,8 @@ imraa_init()
         return MRAA_ERROR_NO_RESOURCES;
     }
 
-    plat->chardev_capable = mraa_is_platform_chardev_interface_capable();
-    if (plat->chardev_capable) {
+    plat->chardev_ver = mraa_check_chardev_ver();
+    if (plat->chardev_ver) {
         syslog(LOG_NOTICE, "gpio: support for chardev interface is activated");
     }
 
diff --git a/src/x86/adlink-ipi.c b/src/x86/adlink-ipi.c
index 849c463d3388..858f1e5f1e63 100644
--- a/src/x86/adlink-ipi.c
+++ b/src/x86/adlink-ipi.c
@@ -860,7 +860,7 @@ mraa_board_t* mraa_lec_al_board()
 
 	b->phy_pin_count = MRAA_LEC_AL_PINCOUNT;
 	b->gpio_count = MRAA_LEC_AL_GPIOCOUNT;
-	b->chardev_capable = 0;
+	b->chardev_ver = 0;
 
 	b->pins = (mraa_pininfo_t*) malloc(sizeof(mraa_pininfo_t) * MRAA_LEC_AL_PINCOUNT);
 	if (b->pins == NULL) {
diff --git a/src/x86/iei_tank.c b/src/x86/iei_tank.c
index 0eee4b723bf0..a2117ef6ce47 100644
--- a/src/x86/iei_tank.c
+++ b/src/x86/iei_tank.c
@@ -26,7 +26,7 @@ mraa_iei_tank()
 
     b->platform_name = PLATFORM_NAME;
     b->phy_pin_count = MRAA_IEI_TANK_PINCOUNT;
-    b->chardev_capable = 1;
+    b->chardev_ver = 1;
 
     b->adv_func = (mraa_adv_func_t*) calloc(1, sizeof(mraa_adv_func_t));
     if (b->adv_func == NULL) {
diff --git a/src/x86/intel_ilk.c b/src/x86/intel_ilk.c
index e71817d91535..dc9409bd5782 100644
--- a/src/x86/intel_ilk.c
+++ b/src/x86/intel_ilk.c
@@ -48,7 +48,7 @@ mraa_intel_ilk()
         goto error;
     }
 
-    b->chardev_capable = 1;
+    b->chardev_ver = 1;
 
     // UART-1
     b->uart_dev_count = 1;
diff --git a/src/x86/intel_joule_expansion.c b/src/x86/intel_joule_expansion.c
index 99d747716898..92c0067ef269 100644
--- a/src/x86/intel_joule_expansion.c
+++ b/src/x86/intel_joule_expansion.c
@@ -26,7 +26,7 @@ mraa_joule_expansion_board()
 
     b->platform_name = PLATFORM_NAME;
     b->phy_pin_count = MRAA_INTEL_JOULE_EXPANSION_PINCOUNT;
-    b->chardev_capable = 1;
+    b->chardev_ver = 1;
     b->aio_count = 0;
     b->adc_raw = 0;
     b->adc_supported = 0;
diff --git a/src/x86/intel_minnow_byt_compatible.c b/src/x86/intel_minnow_byt_compatible.c
index 6b61cf22d50a..dc6867842295 100644
--- a/src/x86/intel_minnow_byt_compatible.c
+++ b/src/x86/intel_minnow_byt_compatible.c
@@ -87,7 +87,7 @@ mraa_intel_minnowboard_byt_compatible(mraa_boolean_t turbot)
         b->platform_version = "Ax";
         b->gpio_count = b->phy_pin_count = MRAA_INTEL_MINNOW_MAX_PINCOUNT;
     }
-    b->chardev_capable = 1;
+    b->chardev_ver = 1;
 
     b->pins = (mraa_pininfo_t*) calloc((size_t) b->phy_pin_count, sizeof(mraa_pininfo_t));
     if (b->pins == NULL) {
diff --git a/src/x86/up2.c b/src/x86/up2.c
index f27068723f83..347438a616ed 100644
--- a/src/x86/up2.c
+++ b/src/x86/up2.c
@@ -101,7 +101,7 @@ mraa_up2_board()
     b->platform_version = PLATFORM_VERSION;
     b->phy_pin_count = MRAA_UP2_PINCOUNT;
     b->gpio_count = MRAA_UP2_GPIOCOUNT;
-    b->chardev_capable = 1;
+    b->chardev_ver = 1;
 
     b->pins = (mraa_pininfo_t*) malloc(sizeof(mraa_pininfo_t) * MRAA_UP2_PINCOUNT);
     if (b->pins == NULL) {
diff --git a/src/x86/up_xtreme.c b/src/x86/up_xtreme.c
index f1408f9c00bb..8acc4afce263 100644
--- a/src/x86/up_xtreme.c
+++ b/src/x86/up_xtreme.c
@@ -120,7 +120,7 @@ mraa_upxtreme_board()
     b->platform_version = PLATFORM_VERSION;
     b->phy_pin_count = MRAA_UPXTREME_PINCOUNT;
     b->gpio_count = MRAA_UPXTREME_GPIOCOUNT;
-    b->chardev_capable = 0;
+    b->chardev_ver = 0;
 
     b->pins = (mraa_pininfo_t*) malloc(sizeof(mraa_pininfo_t) * MRAA_UPXTREME_PINCOUNT);
     if (b->pins == NULL) {
